docs/arch/adr-001.md

```diff
@@ -0,0 +1,38 @@
+# ADR-1: Code Generation
+
+## Context
+
+Serializing and deserializing JSON in Java works best when performing it to/from POJOs.
+While it is possible to serialize into generic `JsonNode` or `Map` objects this results in untyped values that
+must be referenced via String keys, meaning the compiler is limited in what it can check for us.
+
+The classes required for JSON serialization and deserialization are reptitive and error prone to write by hand.
+[Immutables](https://immutables.github.io/) helps to some extent, but there are still many classes and fields to be written.
+
+The [Discord API documentation](https://discordapp.com/developers/docs/intro) is mostly consistent with how it documents
+the JSON structures it returns and accepts.
+
+Code generation is often difficult to use with IDEs.
+Code is often referring to classes that will not be committed, so have to be generated by the IDE.
```

Comment:
`so have to be generated by the IDE` no subject, kind of an awkward sentence. Maybe think `so it must be generated by the IDE in real time to compensate for that` or something similar to kind of show that it's needed to be done in real time rather than just accepting it at compilation.